### DISCUSSION TITLE
How to include built file in serverless function?

### DISCUSSION THREAD

**Discussion by yaquawa** [vercel/vercel#6393] [Help] [answered]
Hi, anyone know how to include built file in serverless function(nodejs runtime)?

I have a `build` command which creates `/dist/index-template.html` under the project root, whenever I push to the git repo, this command will be called to produce my built files.

And I have a serverless function `api/index.js` which want to read the content of the build file `/dist/index-template.html`.

```js
const indexHtml = readFileSync(path.join(__dirname, '../dist/index-template.html'), 'utf8')
```

but this will get an error complaining there is no such a file...
anyone knows how to fix this?

**Comment by mjarkk** [comment]
Have the same problem in [vite-plugin-ssr #98](https://github.com/brillout/vite-plugin-ssr/issues/98).  
My current ~~solution~~ hack is to not build the project on Vercel but do that instead locally and then to not git ignore the build files but this is a very dirty solution.

**Reply by yaquawa** [reply]
yeah, I end up with the same hack..

**Reply by barelyhuman** [reply]
one way is to set function level assets which you can read more about here https://vercel.com/docs/configuration#project/functions for the includeFiles property. 

it doesn't work for all runtimes so you might wanna test it out. 

**example:**

Directory Structure
```
-api
   |-hello.js
-public
   |-template.html 
```

`vercel.json`
```jsonc
{
  "functions": {
    "api/hello.js": {
      includeFiles:["public/**"],
    }
  }
}
```

Then in the actual file you can attempt to access it by using.
`hello.js`
```js
const template = fs.readFileSync(path.join(__dirname,"public","template.html"));
```

make sure you notice the `__dirname` in the path , also do make sure you test this out using `vercel dev` before you actually deploy it.

Hope that helps.

**Reply by brillout** [reply]
@barelyhuman While that works for HTML files, it doesn't work if you are trying to require a JS file for the server (which is the case for vite-plugin-ssr) (the JS file in `public/` would also need to use that hack; you'd need to rewrite all JS files in `public/`.)

**Reply by barelyhuman** [reply]
You can actually handle it with a bundler as a pre setup instead, since it's not the scope of the actual runtime, we can obviously open an issue on the node runtime for the same but if its something you need immediately you can have like a prebuild step for the actual files by inline bundling with rollup for this.

**Reply by brillout** [reply]
@barelyhuman I tried your suggestion to use `includeFiles` but it seems to work only for files commited to Git?

Reproduction:
 1. Deploy https://github.com/brillout/vercel-test.
 2. Go to `/some-non-root-url`.
 3. Watch the function logs and see that `build/` is not included even though `vercel.json#functions['api/index.js'].includeFiles` is set to `build/**`. Note how `package.json#scripts.build` generates `build/someFile.js`.

**Reply by barelyhuman** [reply]
@brillout  yes each action or serverless function will be executed on a single instance so the committed files are what will be available, dynamically built files won't be available on the final instance. 

This is why I was suggesting you create a bundle of the actual function before push using either a precommit hook or a github action that generates them and pushes them into the repo for you. 

This is not ideal but this is how it works right now, as for why it's handled this way, someone from the @vercel-infra  can help you understand that.

**Reply by brillout** [reply]
@barelyhuman There seem to be a way though: I can deploy a Nuxt SSR app without having to commit the build. For example: https://github.com/brillout/vercel-nuxt

**Reply by barelyhuman** [reply]
@brillout are you talking about the dist directory getting added or is your custom build getting added as well?

**Reply by brillout** [reply]
@barelyhuman

> This is why I was suggesting you create a bundle of the actual function before push using either a precommit hook or a github action that generates them and pushes them into the repo for you.

So basically, if I'm understanding you correctly, you are suggesting to commit build files to the repo.

Yet the entire repo https://github.com/brillout/vercel-nuxt doesn't contain any build file. But Nuxt does generate build files that are used by Vercel functions. So I'm wondering how these Vercel functions access these build files that are not committed to the repo.

**Reply by barelyhuman** [reply]
@brillout  to clarify, by build files are you talking about the dist folder generated by nuxt?

**Reply by brillout** [reply]
I'm guessing that the `dist/` folder generated by Nuxt only contains the static files.

But for SSR there must be some other build files (I guess not located in `dist/`).

The repository (https://github.com/brillout/vercel-nuxt) doesn't contain any Vercel function that server-side renders the app.

So I guess it's `@nuxtjs/vercel-builder` that generates/provides the Vercel function that does server-side rendering.

```json5
// vercel.json
{
  "version": 2,
  "builds": [
    {
      "src": "nuxt.config.js",
      "use": "@nuxtjs/vercel-builder",
      "config": {
        "serverFiles": ["server-middleware/**"]
      }
    }
  ]
}
```

I would just dig into creating my own builder but I'm seeing that `vercel.json#builds` is deprecated.

**Reply by barelyhuman** [reply]
So , based on my current knowledge, I know that the functions built by next and nuxt are both executed as needed and while they are generated at build time , the same should be allowed for others as well, but most people I've seen with similar issues were unable to get access to dynamically built files, there's also issues that the same people have raised, you can look them up. 


The solutions or more precisely hacks are that worked involved the built files in the repo. 
I'll take a look at both your examples again after work today. Will let you know what changes can be done to make it work.

**Reply by brillout** [reply]
That'd be awesome. I mean, if Nuxt can do it, then I should be able to do it as well. I'm fine creating my own Vercel builder but I'm worried about the deprecation notice https://vercel.com/docs/configuration#project/builds

**Reply by brillout** [reply]
Even more interesting is SvelteKit: https://github.com/brillout/vercel-svelte.

Vercel is automaticallly picking up the build directory `.vercel_build_output` *including* the serverless functions.

Vercel build logs:

```
12:37:00.840 | Detected Serverless Functions in ".vercel_build_output"
12:37:00.840 | Detected Static Assets in ".vercel_build_output"
12:37:00.840 | Detected Routes Configuration in ".vercel_build_output"
```

And, again, none of the build files are committed in the repo (https://github.com/brillout/vercel-svelte). You can see that `.vercel_build_output` is not there.

**Reply by barelyhuman** [reply]
Oh that, that you can just set via, out directory in the project settings, these all templates have them predefined, I thought you were talking about the custom cases where even the serverless functions are built via something. 

That you can add as an out/build folder in the project settings => general => override out folder and build command and that should be available via the browser but I doubt you can still access their files. 

To see what's accessible you can just open 

Project => View Function Logs => Source (right beside functions) and then => Output

You'll be able to see all the files and folders that the functions can see.

![Screenshot 2021-07-15 at 4 23 30 PM](https://user-images.githubusercontent.com/43572006/125776753-d2ea3a06-7c03-49ae-b3c4-65b92541b509.png)

**Reply by brillout** [reply]
>  I thought you were talking about the custom cases where even the serverless functions are built via something.

That is actually exactly what I'm talking about.

Vercel production files:

![Screenshot 2021-07-15 13 20 57](https://user-images.githubusercontent.com/1005638/125780660-f6b4ff0b-1ad5-48e3-bee0-d25467aea51c.png)

Note the serverless function `.vercel/functions/render/index`.

Repository files:

![Screenshot 2021-07-15 13 22 27](https://user-images.githubusercontent.com/1005638/125780666-2a9f2668-0d19-4a68-8a22-d3ba93fd7de6.png)

There are no sererless functions defined in there, in particular `.vercel/functions/render/index` is not there.

Vercel config:

![Screenshot 2021-07-15 13 25 13](https://user-images.githubusercontent.com/1005638/125780524-a68113d5-6347-4981-a84b-ca3cad127ade.png)

So the interesting thing here is that the serverless function is not committed to the repo.

When I run `npm run build` locally it generates the directory `.vercel_build_output` which does contain the serverless function at `.vercel_build_output/functions/node/render/index.js`.

So what's going on here?

**Reply by barelyhuman** [reply]
Well you can kinda replicate the same, that was a part of the answer I gave.

> You can actually handle it with a bundler as a pre setup instead, since it's not the scope of the actual runtime, we can obviously open an issue on the node runtime for the same but if its something you need immediately you can have like a prebuild step for the actual files by inline bundling with rollup for this.

What I mean was that you take something like rollup, configure it to build your functions and land them in the api folder with the needed function, that will be copied around to each instance and you'll be able to invoke these. 

I'll create an example repo of what I mean and then we can see if we are on the same page or still expecting 2 different things.

**Reply by brillout** [reply]
>  you take something like rollup, configure it to build your functions and land them in the api folder

But then I would need to commit these generated files to the repo, correct?

The interesting thing about the example I was showcasing is that the serverless function is *not* committed.

**Reply by barelyhuman** [reply]
no no no, you don't have to commit them, I'll be done with work in like an hour and I'll send through a repo of what I'm talking about.

**Reply by brillout** [reply]
Neat... looking forward to see how to do it.

**Reply by barelyhuman** [reply]
Repo: https://github.com/barelyhuman/api-build-test

Requests you can make 
- https://api-build-test.vercel.app/api/home?name=Sid
- https://api-build-test.vercel.app/api/home

Both being handled by a helper generated at runtime.
I hope that's what you wanted, if not, let me know

**Reply by brillout** [reply]
The script `npm run vercel-build` generates the file `api_helpers/custom-function.js` but it is already committed.

![Screenshot 2021-07-15 14 50 33](https://user-images.githubusercontent.com/1005638/125791121-21545a28-d4a3-4121-9bb6-7315108bcb4b.png)

My goal is that it is not committed.

**Reply by barelyhuman** [reply]
Latest commit added that to the ignore as well, if you wish, I'll force ammend the ignored file, it still works for me, could you confirm as well?

**Reply by brillout** [reply]
Ah neat, let me try

**Reply by brillout** [reply]
It seems to work... nice.

So why is `api_helpers/` added? Why did my previous attemps fail but it works now?

**Reply by brillout** [reply]
Does Vercel treat `api_helpers/` differently than other directories?

**Reply by barelyhuman** [reply]
so the concept is that you might need helpers or other things that can be heavy to add into the repo but is still needed at run time and next.js kinda does this, I'm still not sure how svelte does it.
You can use the script `vercel-build` for that. 
This will build post the normal build command and is only available in the JS projects sadly, I tried it in a go project and it burst of my face. 

No the name has no significance,  the package json script does. Anything generated by that is kept on the instance. 

If you wanna test this out locally,  you can use the `vercel dev` command. The command will keep generating the file so don't freak out, that's normal. 

Though, I did assume you wanted the API's to be generated dynamically, which sadly I couldn't achieve , my project wouldn't get recognized at all so I ended up giving this as an example. 

You can read more about it 
https://vercel.com/docs/runtimes#advanced-usage/advanced-node-js-usage/custom-build-step-for-node-js

**Reply by brillout** [reply]
That's fine, that's good enough for my purposes.

Super excited to try it now for my use case. I'll keep you updated.

Thousand thanks in the meantime!

**Reply by barelyhuman** [reply]
Glad I could help :)

**Reply by brillout** [reply]
@barelyhuman

It works https://vercel-vite-ssr-test-pixahavqg-vercel-brilloutcom.vercel.app/ 👌.

> Glad I could help :)

You also helped many of my users who wanted to use Vercel https://vite-plugin-ssr.com/vercel :-).

Two weird things though:
 - If `package.json#scripts.build` is not defined then Vercel ignores the static files directory.
 - Vercel runs `package.json#scripts.vercel-build` twice. No clue why.

**Reply by barelyhuman** [reply]
@brillout  I saw it run twice as well but that was when I had a package.json in the `api` directory, i'm not sure why it's happening for you but I'll go through the vercel source code to see when and how the vercel build is triggered, my guess still remains that it runs it for each package.json it finds.

**Reply by barelyhuman** [reply]
> if package.json#scripts.build is not defined then Vercel ignores the static files directory.

That is either a bug or a feature, something that vercel team can answer, though I saw that happen too and hence the empty build script in the test repository I gave yesterday

**Reply by brillout** [reply]
👍

About the pkg thing: if you look at the source https://github.com/brillout/vercel-vite-ssr-test there is only one `package.json` so seems like it's something else.

Keep me updated about the bugs; It'd be nice to remove the hacks from the Vercel + vite-plugin-ssr example.

Thanks again!

**Reply by barelyhuman** [reply]
@brillout 
Sure, no probs :)

**Comment by barelyhuman** [comment, marked as answer]
@yaquawa In case you don't want to read the whole conversation, 
the answers are
- using Github Actions to commit generated files post push 
- using includeFiles 
- using `vercel-build` in package.json to generate files during vercel portal build 

Hope one of them works for you :)

**Reply by yaquawa** [reply]
Thanks @barelyhuman , but that won't work for my case, because I'm trying to include a file that was created after the "build command" but not the `vercel-build`. I guess I have no other workaround for my case.

**Reply by barelyhuman** [reply]
@yaquawa 
Yeah, so you move the generation of the html file to `vercel-build`, that should work for you, 
https://github.com/barelyhuman/api-build-test
Check the above repo for the same use-case.

**Reply by yaquawa** [reply]
Thanks @barelyhuman I got the idea👍🏼
But that means I have to build the source code twice before I can deploy it, one for the "build" one for the "vercel-build", which takes too many time for me. I guess I'll stick with the tricky workaround for now..

**Reply by barelyhuman** [reply]
@yaquawa  build twice? but both build and vercel-build will be run by the vercel portal when you push, you don't have to manually run it even once. 

Sorry, I'm trying to understand why it's going to be have to run twice.

**Reply by yaquawa** [reply]
@barelyhuman Yes, I understand that it'll be run automatically. The file I want to include is a file that created via "build", so I have to run the same "build" for "vercel-build", 

I mean:
```json
{
  "build": "build-my-project-cmd",
  "vercel-build":  "build-my-project-cmd"
}
```

**Reply by barelyhuman** [reply]
Oh okay, yeah you don't need to have 2 builds if they are doing the same thing. 
I'll try to simplify , but if you already understood this sorry to repeat. 

1. You will use `vercel-build` for the dynamically generated files, you can leave the `build` command to be empty with a `echo "ran build"` type of command. 
2. If you have other dependent scripts that run build and that's stopping you from running this, just add a `prebuild` script that runs `vercel-build` for you while working locally and `build` can still be empty. 

Hope that helps :)

**Reply by yaquawa** [reply]
Thanks @barelyhuman , I'm not sure if the files built from "vercel-build" have the same scope that "build" has, if it is, this will be the solution.

**Reply by barelyhuman** [reply]
@yaquawa by scope if you mean they have access to the same folders, then yes they do, either way, try it out, if it works out , amazing.

**Reply by yaquawa** [reply]
@barelyhuman Thanks! I'll try it later on.
